File: CSharp/Bank_Account_and_Transaction_Ledger/Services/Ledger.cs
using Models;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Services
{
    public class Ledger
    {
        private Dictionary<string, Account> accounts = new Dictionary<string, Account>();
        private List<Transaction> transactions = new List<Transaction>();
        private string filepath;

        public Ledger(string filepath)
        {
            this.filepath = filepath;
        }

        public bool AddAccount(Account account)
        {
            if (accounts.ContainsKey(account.AccountNumber))
            {
                Console.WriteLine($"Account {account.AccountNumber} already exists.");
                return false;
            }
            accounts[account.AccountNumber] = account;
            return true;
        }

        public void Deposit(string accountNumber, double amount)
        {
            if (!accounts.ContainsKey(accountNumber))
            {
                Console.WriteLine($"No account found with number {accountNumber}");
                return;
            }
            Account account = accounts[accountNumber];
            account.Deposit(amount);
            Transaction transaction = new Transaction(accountNumber, "deposit", amount);
            transactions.Add(transaction);
            TransactionStorage.AppendTransaction(transaction, filepath);
        }

        public void Withdraw(string accountNumber, double amount)
        {
            if (!accounts.ContainsKey(accountNumber))
            {
                Console.WriteLine($"No account found with number {accountNumber}");
                return;
            }
            Account account = accounts[accountNumber];
            account.Withdraw(amount);
            Transaction transaction = new Transaction(accountNumber, "withdraw", amount);
            transactions.Add(transaction);
            TransactionStorage.AppendTransaction(transaction, filepath);
        }

        public void ApplyMonthlyUpdateAll()
        {
            foreach(Account account in accounts.Values)
            {
                account.ApplyMonthlyUpdate();
            }
        }

        public List<Transaction> GetAllTransactions()
        {
            return transactions;
        }

        public Account GetAccount(string accountNumber)
        {
            return accounts.ContainsKey(accountNumber) ? accounts[accountNumber] : null;
        }
    }
}

File: Bank_Account_and_Transaction_Ledger/Services/ledger.py
from models.transaction import Transaction
from Services.storage import append_transaction

class Ledger:
    def __init__(self, filepath):
        self.accounts = {}
        self.transactions = []
        self.filepath = filepath
        
    def add_account(self, account):
        if account.account_number in self.accounts:
            print(f"Account {account.account_number} already exists.")
            return False
        self.accounts[account.account_number] = account
        return True
    
    def deposit(self, account_number, amount):
        if account_number not in self.accounts:
            print(f"No account found wiht number {account_number}")
            return
        account = self.accounts[account_number]
        account.deposit(amount)
        transaction = Transaction(account_number, "deposit", amount)
        self.transactions.append(transaction)
        append_transaction(transaction, self.filepath)
        
    def withdraw(self, account_number, amount):
        if account_number not in self.accounts:
            print(f"No account found wiht number {account_number}")
            return
        account = self.accounts[account_number]
        account.withdraw(amount)
        transaction = Transaction(account_number, "withdraw", amount)
        self.transactions.append(transaction)
        append_transaction(transaction, self.filepath)
        
    def apply_monthly_update_all(self):
        for account in self.accounts.values():
            account.apply_monthly_update()
            
    def get_all_transactions(self):
        return self.transactions
    
    def get_account(self, account_number):
        return self.accounts.get(account_number)